### ISSUE TITLE
`prepareParameters` can break nested arrays

### ISSUE THREAD

**Issue by thorlucas** [cartalyst/stripe#180]
Hi,

I noticed a discrepancy between the official Stripe package and this one when creating a Checkout session with multiple line items:

Official stripe package:

```php
\Stripe\Stripe::setApiKey('sk_test_xxxxx');
$session = \Stripe\Checkout\Session::create([	
	'payment_method_types' => ['card'],
	'line_items' => [
		[
			'price_data' => [
				'currency' => 'usd',
				'product_data' => [
					'name' => 'T-shirt',
				],
				'unit_amount' => 2000,
			],
		'quantity' => 1,
		],
		[
			'price_data' => [
				'currency' => 'usd',
				'product_data' => [
					'name' => 'T-shirt',
				],
				'unit_amount' => 2000,
			],
		'quantity' => 1,
		],
	],
	'mode' => 'payment',
	'success_url' => 'https://example.com/success',
	'cancel_url' => 'https://example.com/cancel',
]);
// Works, returns the working session object
```

Cartalyst Stripe:
```php
$session = Stripe::checkout()->sessions()->create([	
	'payment_method_types' => ['card'],
	'line_items' => [
		[
			'price_data' => [
				'currency' => 'usd',
				'product_data' => [
					'name' => 'T-shirt',
				],
				'unit_amount' => 2000,
			],
		'quantity' => 1,
		],
		[
			'price_data' => [
				'currency' => 'usd',
				'product_data' => [
					'name' => 'T-shirt',
				],
				'unit_amount' => 2000,
			],
		'quantity' => 1,
		],
	],
	'mode' => 'payment',
	'success_url' => 'https://example.com/success',
	'cancel_url' => 'https://example.com/cancel',
]);
// Cartalyst/Stripe/Exception/MissingParameterException with message 'You must provide either price or price_data for each line item when using prices.'
```

**Comment by GrahamCampbell** [comment]
@thorlucas Can you find out the raw http request body which is sent, so we can compare the difference, in both cases?

**Comment by brunogaspar** [comment]
From a very quick test, the issue comes from [here](https://github.com/cartalyst/stripe/blob/2.4/src/Utility.php#L51)

Unfortunately, i don't have enough time today to look into this.

**Comment by thorlucas** [comment]
Ah yes, it looks like this `preg_replace` converts the input to this:

```
[
     "payment_method_types" => [
       "card",
     ],
     "line_items" => [
       [
         "price_data" => [
           "currency" => "usd",
         ],
       ],
       [
         "price_data" => [
           "product_data" => [
             "name" => "T-shirt",
           ],
         ],
       ],
       [
         "price_data" => [
           "unit_amount" => "2000",
         ],
       ],
       [
         "quantity" => "1",
       ],
       [
         "price_data" => [
           "currency" => "usd",
         ],
       ],
       [
         "price_data" => [
           "product_data" => [
             "name" => "T-shirt",
           ],
         ],
       ],
       [
         "price_data" => [
           "unit_amount" => "2000",
         ],
       ],
       [
         "quantity" => "1",
       ],
     ],
     "mode" => "payment",
     "success_url" => "https://example.com/success",
     "cancel_url" => "https://example.com/cancel",
   ]
```

**Comment by brunogaspar** [comment]
Yes, correct.

I believe this was mostly because of a bug after updating to Guzzle 6 and something on Stripe itself due to not allowing arrays without proper indexes (mostly numeric), need to run the tests locally to ensure nothing breaks.
This was done 5 years ago or something so i don't remember the specifics.

**Comment by GrahamCampbell** [comment]
Perhaps the issue was encoding URI query strings. Is that code shared with body encoding?

**Comment by brunogaspar** [comment]
Not sure.

I tried with

```php
preg_replace('/\%5B\d+\%5D/', '%5B%5D', urlencode (http_build_query($parameters)));
```

And it worked then i tried with just returning the output of the `http_build_query` and also worked, all tests passed, so i assume that the bug/issue is now gone 🤷‍♂️  lol

Going to merge the branch and tag a release tomorrow.

**Comment by brunogaspar** [comment]
Released v2.4.2 which should fix the issue.
